### ISSUE TITLE
Potential risk of deadlock from calling listeners under locks

### ISSUE THREAD

**Issue by zhangkun83** [grpc/grpc-java#1408]
Methods of `ClientTransport.Listener` and `ServerTransportListener` are usually called under a lock. The biggest reason for locking is to guarantee the ordering of multiple methods on the same listener.

However, these listeners usually call into channel layer code, and may in turn acquire locks from there, which forms a `transport lock -> channel lock` lock order. On the other hand, when channel layer calls into transport layer, it's possible to form a `channel lock -> transport lock` lock order, which makes deadlock possible.

It's unlikely an issue today because there is an implicit rule today that channel layer will not hold any lock while calling into transport. However, as the code base grows, it will become harder to keep track of such requirement.

A possible solution is to always schedule listener calls on a serialized executor, with the cost of a thread, so that listener order can be guaranteed without the need of locking. There may be better options.

**Comment by ejona86** [comment]
When driving home yesterday I realized: I think we can use `SerializingExecutor` _in addition to_ `ThreadlessExecutor`.

**Comment by zhangkun83** [comment]
It's just not only transport listeners. Calling stream listeners can also have the deadlock issue, specially in `InProcessTransport`. The deadlock triggered by grpc/grpc-java/pull/1503 is an example.

**Comment by zhangkun83** [comment]
We have established the lock order, that we do not hold any lock while calling into transport., while transport can call callbacks while holding its lock. We have made many changes to escape locks when crossing component boundary. The only cross-component lock allowed today is the transport lock, which exists in OkHttp and InProcess. Since the deadlock risk has reduced dramatically, I am closing this issue.
